File: Data Structures & Algorithms/minimum-stack/submission-10.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min = float("inf")
    def push(self, val: int) -> None:
        if not self.stack:
            self.min = val
            self.stack.append(0)
        else :
            self.stack.append(val - self.min)
            self.min = min(self.min, val)
    def pop(self) -> None:
        if self.stack[-1] < 0:
            self.min = self.min - self.stack[-1]
        self.stack.pop()
    def top(self) -> int:
        if self.stack[-1] > 0:
            return self.stack[-1] + self.min
        else :
            return self.min
    def getMin(self) -> int:
        return self.min

File: Data Structures & Algorithms/valid-binary-search-tree/submission-10.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<tuple<TreeNode*, int, int>> st;
        int min{numeric_limits<int>::min()};
        int max{numeric_limits<int>::max()};
        st.push({root, min, max});

        while (st.size()) {
            auto [node, mini, maxi]{st.top()};
            st.pop();
            if (!node) continue;
            if (node->val <= mini || node->val >= maxi) return false;
            st.push({node->left, mini, node->val});
            st.push({node->right, node->val, maxi});
        }
        return true;
    }
};
